lib/elastic_apm/transport/connection.rb

```diff
@@ -1,204 +1,108 @@
+
 module ElasticAPM
   module Transport
     # @api private
-    class Connection # rubocop:disable Metrics/ClassLength
+    class Connection
       include Logging
 
-      class FailedToConnectError < InternalError; end
-
-      # @api private
-      # HTTP.rb calls #rewind the body stream which IO.pipes don't support
-      class ModdedIO < IO
-        def self.pipe(ext_enc = nil)
-          super(ext_enc).tap do |rw|
-            rw[0].define_singleton_method(:rewind) { nil }
-          end
-        end
-      end
-
       HEADERS = {
         'Content-Type' => 'application/x-ndjson',
         'Transfer-Encoding' => 'chunked'
       }.freeze
-      GZIP_HEADERS = HEADERS.merge('Content-Encoding' => 'gzip').freeze
+      GZIP_HEADERS = HEADERS.merge(
+        'Content-Encoding' => 'gzip'
+      ).freeze
 
-      # rubocop:disable Metrics/MethodLength
       def initialize(config, metadata)
         @config = config
-        @metadata = metadata.to_json
-
+        @metadata = JSON.fast_generate(metadata)
         @url = config.server_url + '/intake/v2/events'
-
-        headers =
-          (@config.http_compression? ? GZIP_HEADERS : HEADERS).dup
-
-        if (token = config.secret_token)
-          headers['Authorization'] = "Bearer #{token}"
-        end
-
-        @client = HTTP.headers(headers).persistent(@url)
-
-        configure_proxy
-        configure_ssl
+        @headers = build_headers
+        @ssl_context = build_ssl_context
 
         @mutex = Mutex.new
       end
-      # rubocop:enable Metrics/MethodLength
-
-      def configure_proxy
-        unless @config.proxy_address && @config.proxy_port
-          return
-        end
-
-        @client = @client.via(
-          @config.proxy_address,
-          @config.proxy_port,
-          @config.proxy_username,
-          @config.proxy_password,
-          @config.proxy_headers
-        )
-      end
-
-      def configure_ssl
-        return unless @config.use_ssl? && @config.server_ca_cert
-
-        @ssl_context = OpenSSL::SSL::SSLContext.new.tap do |context|
-          context.ca_file = @config.server_ca_cert
-        end
-      end
 
+      # rubocop:disable Metrics/MethodLength, Metrics/AbcSize
       def write(str)
-        return if @config.disable_send
+        return false if @config.disable_send
 
-        connect_unless_connected
-
-        @mutex.synchronize { append(str) }
-
-        return unless @bytes_sent >= @config.api_request_size
-
-        flush
-      rescue FailedToConnectError => e
-        error "Couldn't establish connection to APM Server:\n%p", e
-        flush
-
-        nil
-      end
-
-      def connected?
-        @mutex.synchronize { @connected }
-      end
-
-      def flush
-        @mutex.synchronize do
-          return unless @connected
-
-          debug 'Closing request'
-          @wr.close
-          @conn_thread.join 5 if @conn_thread
+        begin
+          w = 0
+          @mutex.synchronize do
```

Comment:
why is there a mutex? doesn't each worker have its own connection?

Reply:
Because there is a separate scheduled thread taking care of closing the connection after a timeout.

Reply:
:ok_hand:

I think it would be useful to have some comments about expected concurrency, to make this clear.